fem/fe/fe_rbf.hpp

```diff
@@ -0,0 +1,556 @@
+                                  DenseMatrix &I) const
+   { CheckScalarFE(fe).ScalarLocalInterpolation(Trans, I, *this); }
+};
+
+// Finite element using base radial basis functions without polynomial corrections
+class RBFFiniteElement : public KernelFiniteElement
+{
+private:
+#ifndef MFEM_THREAD_SAFE
+   mutable double r_scr, df_scr, ddf_scr;
+   mutable Vector x_scr, y_scr, dy_scr, dr_scr;
+   mutable DenseMatrix ddr_scr;
+   mutable int cInd[3][2];
+#endif
+   bool isCompact; // Is the RBF with the given h compact?
+   int dimPoints[3];
+   int numPointsD; // Number of points across the element in each D
+   double delta; // Distance between points
+   double h; // Shape parameter, approx number of points in 1d support radius
+   double hPhys; // Shape parameter times distance between points times HNorm
+   double hPhysInv; // Inverse hPhys
+   double radPhys; // Radius adjusted by h
+   double faceFactor; // 1.0 = points end on face, 0.0 = points end at dx/2 from face
+   const RBFFunction *rbf;
+   const DistanceMetric *distance;
+   void InitializeGeometry();
+
+   // Get the dimensionless distance from x to the center of the RBF indexed i
+   virtual void DistanceVec(const int i,
+                            const Vector &x,
+                            Vector &y) const;
+
+public:
+   RBFFiniteElement(const int D,
+                    const int numPointsD,
+                    const int rbfType,
+                    const int distNorm,
+                    const int intOrder,
+                    const double h,
+                    const double faceFactor);
+   virtual ~RBFFiniteElement() { delete rbf; delete distance; }
+
+   virtual bool TensorIndexed() const { return true; }
+   virtual void GetCompactIndices(const Vector &ip,
+                                  int (&indices)[3][2]) const;
+   virtual void GetGlobalIndices(const Vector &ip,
+                                 int (&indices)[3][2]) const;
+   virtual void GetTensorIndices(const Vector &ip,
+                                 int (&indices)[3][2]) const;
+   virtual void GetTensorNumPoints(int (&tNumPoints)[3]) const
+   {
+      tNumPoints[0] = dimPoints[0];
+      tNumPoints[1] = dimPoints[1];
+      tNumPoints[2] = dimPoints[2];
+   }
+
+   virtual bool IsCompact() const { return isCompact; }
+   virtual const RBFFunction *Kernel() const { return rbf; }
+
+   virtual void CalcShape(const IntegrationPoint &ip,
+                          Vector &shape) const;
+   virtual void CalcDShape(const IntegrationPoint &ip,
+                           DenseMatrix &dshape) const;
+   virtual void CalcHessian(const IntegrationPoint &ip,
+                            DenseMatrix &hess) const;
+};
+
+/* Reproducing kernel finite element, which includes polynomial corrections
+   to the standard radial basis function finite element to guarantee a
+   chosen order of accuracy */
+class RKFiniteElement : public KernelFiniteElement
+{
+private:
+#ifndef MFEM_THREAD_SAFE
```

Comment:
If you care so much about `MFEM_THREAD_SAFE`, why do you bother to have them as members? They're small and can always be created locally, removing all the extra `#ifdef` code.

Reply:
The sizes of the variables range from (dimension) to (dimension * num_poly * num_poly), where num_poly is the number of polynomial correction terms. The most common corrections will probably be linear (in 3D, num_poly = 4, so max size 48) or quadratic (num_poly=10, so max size 300). What do you think a good size cutoff is for something that should be a data member vs a local variable?

Reply:
It's ok then, I didn't realize these could grow so much.